Add SET_FLUID behavior.

murphy-slaw opened the pull request:
Adds a new behavior, SET_FLUID, which will set the cauldron's contents to the fluid specified by an Identifier. Returns PASS if the identifier is invalid.

murphy-slaw committed:
Add SET_FLUID behavior.

Sploder12 commented on README.md:
```diff
@@ -187,8 +187,9 @@ Note: `"SUCCESS"` refers to `ActionResult.CONSUME` when running on the server an
-| `"ADD_LEVEL"`          | v0.3.0+ | Adds a fluid level to the cauldron, will return `"PASS"` when cauldron is full         | `"dilute"`: (Optional = true) Boolean - If dilution should occur. `"fluid"`: (Optional) Identifier - fluid to try to add, will fail if the cauldron already contains a different fluid unless that fluid is `Fluids.EMPTY`                                                        ...
+| `"ADD_LEVEL"`          | v0.3.0+ | Adds a fluid level to the cauldron, will return `"PASS"` when cauldron is full         | `"dilute"`: (Optional = true) Boolean - If dilution should occur. `"fluid"`: (Optional) Identifier - fluid to try to add, will fail if the cauldron already contains a different fluid unless that fluid is `Fluids.EMPTY`                                                        ...
 | `"REMOVE_LEVEL"`       | v0.3.0+ | Removes a fluid level from the cauldron, will return `"PASS"` when cauldron is empty   |                                                                                                                                                                                                                                                                                   ...
+| `"SET_FLUID"`          | v0.5.0+ | Sets the fluid contained by the cauldron, will return `"PASS"` if the fluid is invalid | `"fluid"`: Identifier - fluid to set                                                                                                                                                                                                                                              ...
```
Since `v0.5.1` is already released, the version for `"SET_FLUID"` should be something else. Not sure what that something else is, but a placeholder like `unreleased` would work.

murphy-slaw replied:
Done.

Sploder12 commented on src/main/java/net/sploder12/potioncraft/meta/templates/Effect.java:
```diff
@@ -244,4 +244,16 @@ public interface Effect {
         };
     };
+
+    MetaEffectTemplate SET_FLUID = (params, file) -> {
```
Could you add a small comment similar to the other effects please?

murphy-slaw replied:
Done.

Sploder12 reviewed (changes requested):
Hi @murphy-slaw, thank you for your contribution!

While testing your changes I found an issue where the fluid does not get properly updated if you use `SET_FLUID` on a vanilla cauldron. The issue seems to stem from how `.../meta/CauldronData.java:transformBlock` handles updates to vanilla cauldrons, currently it only updates them if the level changed. Adding a check for fluid change should fix it. 

Also, what do you think of logging a warning if `"fluid"` is invalid?

murphy-slaw committed:
Tweak transformBlock to check for fluid changes.
Add requested comments and doc changes
Tweak PotionCauldronRenderer to not ignore fluid tints

murphy-slaw committed:
Log warning on invalid fluid id in SET_FLUID

murphy-slaw commented:
One additional change I made was to use the fluid's color/tint if there are no effects. KubeJS uses grayscale textures with tints for custom fluids unless the user specifies a texture, and these were showing up as white in the cauldron.

My current implementation seems kind of janky, if you can think a better approach I'd be glad to hear it.

murphy-slaw committed:
Fix edge case where the potion cauldron is the default cauldron for a custom fluid but there are no effects applied

Sploder12 reviewed (approved):
Thank you again @murphy-slaw, everything looks good to me! 

Your solution to the tint problem seems fine to me. It's a small change so even if I had a better solution it wouldn't be a hard adjust.
